File: PDBData/loadDB.pl
#!/usr/bin/perl
use DBI;
use strict;
#
require "bdconn.pl";
#
my $dbh=connectDB();
# Clean Tables !!
foreach my $tab ('entry', 'source', 'sequence', 'author', 'author_has_entry', 'expType', 'comptype', 'entry_has_source', 'expClasse') {
   print "Cleaning $tab\n";
   $dbh->do ("DELETE FROM $tab");
}
foreach my $tab ('source','author','expType','comptype','expClasse') {
   $dbh->do ("ALTER TABLE $tab AUTO_INCREMENT=1");
}
#
$dbh->do("SET FOREIGN_KEY_CHECKS=0");
#
print "Authors...";
my %AUTHORS;
my %author_has_entry;
my $sthAuthor=$dbh->prepare ("INSERT INTO author (author) VALUES (?)");
my $sthEntryAuthor=$dbh->prepare("INSERT INTO author_has_entry VALUES (?,?)");
open AUTS, "author.idx";
while (<AUTS>) {
  next if !/;/;
  chomp;
  my ($idCode, $author) = split / *; */;
  next if (!$author);
  if ($author && !$AUTHORS{$author}) {
  	$sthAuthor->execute ($author);
        $AUTHORS{$author}=$dbh->last_insert_id('','','Author','idAuthor');
  }
  if (!$author_has_entry{"$AUTHORS{$author}-$idCode"}) {
     $sthEntryAuthor->execute ($AUTHORS{$author},$idCode);
     $author_has_entry{"$AUTHORS{$author}-$idCode"}=1;
  }
  print ".";
}
close AUTS;
print "ok\n";
#
#
print "Sources...";
my %SOURCES;
my %Entry_has_source;
my $sthSource =$dbh->prepare("INSERT INTO source (source) VALUES (?)");
my $sthEntrySource = $dbh->prepare("INSERT INTO entry_has_source (idCode,idSource) VALUES (?,?)");
open SOUR, "source.idx";
while (<SOUR>) {
   chomp;
   my ($idCode,$source) = split ' ', $_, 2;
   next if (!$source) || (length($idCode) != 4);
   foreach my $s (split /; */, $source) {
      if (!$SOURCES{$s}) {
         $sthSource->execute ($s);
	 $SOURCES{$s}=$dbh->last_insert_id('','','source','idSource');
      }
      $sthEntrySource->execute($idCode, $SOURCES{$s});
   }
   print ".";
}
close SOUR;
print "ok\n";
#
#
print "Entries...";
open ENTR, "entries.idx";
my $sthEntry = $dbh->prepare ("INSERT INTO entry (idCode, header, ascessionDate, compound, resolution) VALUES (?,?,?,?,?)");
my %ExpTypes;
my $sthExpType = $dbh->prepare ("INSERT INTO expType (ExpType) VALUES (?)");
my $sthEntryExpType = $dbh->prepare ("UPDATE entry SET idExpType=? WHERE idCode=?");
my %expTypesbyCode;
while (<ENTR>) {
   chomp;
   my ($idCode, $header, $ascDate, $compound, $source, $authorList, $resol, $expType) = split /\t/;
   next if (length($idCode) != 4);
   if ($resol =~/NOT/) {
	$resol = 0
   }
   if ($resol =~ /,/) {
	my @r = split /,/, $resol;
	$resol =$r[0];
   }
   $compound = substr($compound,0,255);
   $sthEntry-> execute ($idCode,$header, $ascDate, $compound, $resol);   	
	if (!$ExpTypes{$expType }) {
		$sthExpType->execute($expType);
		$ExpTypes{$expType}=$dbh->last_insert_id('','','ExpType','idExpType');
	}
	$sthEntryExpType->execute($ExpTypes{$expType},$idCode);
	$expTypesbyCode{$idCode}=$expType;
   print ".";
}
close ENTR;
print "ok\n";
#
#
open EXPCL , 'pdb_entry_type.txt';
my %expClasses;
my %compTypes;
my $sthExpClasse=$dbh->prepare('INSERT INTO expClasse (expClasse) VALUES (?)');
my $sthcompType=$dbh->prepare('INSERT INTO comptype (type) VALUES (?)');
my $entryUpdate= $dbh->prepare('UPDATE entry SET idCompType=? WHERE idCode=?');
my $expTypeUpdate=$dbh->prepare('UPDATE expType SET idExpClasse=? where ExpType=?');

while (<EXPCL>) {
   chomp;
	my ($idCode, $compType, $expClass) = split ' ';
	$idCode =~ tr/a-z/A-Z/;
	if (!$expClasses{$expClass}) {
		$sthExpClasse->execute($expClass);
		$expClasses{$expClass}=$dbh->last_insert_id('','','expClasses','idExpClass');
	}
	if (!$compTypes{$compType}) {
		$sthcompType->execute($compType);
		$compTypes{$compType}=$dbh->last_insert_id('','','compType','idCompType');
	}
	$expTypeUpdate->execute($expClasses{$expClass}, $expTypesbyCode{$idCode});
	$entryUpdate->execute($compTypes{$compType},$idCode);
}	
#Sequence
print "Sequences...";
my $codes=[];
my %CODES;
my $sthInsert= $dbh->prepare ("INSERT INTO sequence (idCode,chain,sequence,header) VALUES (?,?,?,?)");
open "SEQS", "pdb_seqres.txt";
my $seq;
my $idPdb;
my $chain;
my $header;
while (<SEQS>) {
   chomp;
   if (/^>/) {
	if ($seq) {
		$seq =~ s/\n//g;
		$idPdb =~ tr/a-z/A-Z/;
		$chain =~ s/ //;
   		$sthInsert->execute ($idPdb,$chain,$seq, $header) || die $DBI::errstr;
	        $seq="";
	}
	/^>([^_]*)_(.*)mol:(\S*) length:(\S*) (.*)/;
	($idPdb,$chain,$header)=($1,$2,"$1 $2 $3 $4 $5");
	print "$header\n";
   }
   else {$seq .= $_};
};
print "ok\n";
disconnectDB($dbh);

File: PDBData/scripts/loadDB.py
import pymysql
import re

#Database connection
database = "pdb";
host = "localhost";
user="dbw00"; 
passwd = "dbw2018";

connection = pymysql.connect(host='localhost',
                user=user,
                password=passwd,
                db=database,
                charset='utf8mb4',
                autocommit=True
            )

## Turn off FKs
connection.cursor().execute("SET FOREIGN_KEY_CHECKS=0")

## Clean Tables !!
for tab in (
    'entry', 'source', 'sequence', 'author', 'author_has_entry', 
    'expType', 'comptype', 'entry_has_source', 'expClasse'):
    try:
        print("Cleaning {}".format(tab))
        connection.cursor().execute("DELETE FROM "+ tab)
    except OSError as e:
        sys.exit(e.msg)
    
        
for tab in  ('source','author','expType','comptype','expClasse'):
    connection.cursor().execute("ALTER TABLE " + tab + " AUTO_INCREMENT=1")
print("Authors...")
AUTHORS = {}
author_has_entry = {}
sthAuthor = "INSERT INTO author (author) VALUES (%s)"
sthEntryAuthor = "INSERT INTO author_has_entry VALUES (%s,%s)"
with open('author.idx', 'r') as AUTS:
    for line in AUTS:
        line = line.rstrip()
        if ' ; ' in line:
            idCode, author = line.split(" ; ")
            if not idCode or not author:
                continue
            if author and author not in AUTHORS: 
                with connection.cursor() as c:
                    c.execute(sthAuthor, author)
                    author_id = c.lastrowid                    
                    AUTHORS[author]=author_id
                    aek = "{}{}".format(author_id, idCode)
                    if aek not in author_has_entry: 
                        c.execute(sthEntryAuthor, (author_id, idCode))
                        author_has_entry[aek] = True
print("ok")
##
##
print ("Sources...")
SOURCES = {}
sthSource = "INSERT INTO source (source) VALUES (%s)"
sthEntrySource = "INSERT INTO entry_has_source (idCode,idSource) VALUES (%s, %s)"
with open ("source.idx", 'r') as SOUR:
    for line in SOUR:
        line = line.rstrip()
        if ' ' not in line:
            continue
        idCode, source  = line.split(maxsplit=1)
        if not source or len(idCode) != 4:
            continue
        for s in source.split('; '):
            with connection.cursor() as c:
                if s not in SOURCES:
                    c.execute(sthSource, s)
                    SOURCES[s] = c.lastrowid
                c.execute(sthEntrySource, (idCode, SOURCES[s]))
print("ok")
##
##
print("Entries...")

sthEntry = "INSERT INTO entry (idCode, header, ascessionDate, compound, resolution) VALUES (%s, %s, %s, %s, %s)"
ExpTypes = {}
sthExpType = "INSERT INTO expType (ExpType) VALUES (%s)"
sthEntryExpType = "UPDATE entry SET idExpType=%s WHERE idCode=%s"
expTypesbyCode = {}

with open ("entries.idx", 'r') as ENTR:
    for line in ENTR:
        line = line.rstrip()
        if "\t" not in line:
            continue
        idCode, header, ascDate, compound, source, authorList, resol, expType = line.split("\t")
        if len(idCode) != 4:
            continue
        if 'NOT' in resol:
            resol = 0
        if ',' in str(resol):
            r = resol.split(",")
            resol = r[0]
        compound = compound[:255]
        with connection.cursor() as c:
            c.execute(sthEntry, (idCode, header, ascDate, compound, resol))
            if expType not in ExpTypes:
                c.execute(sthExpType, expType)
                ExpTypes[expType] = c.lastrowid
            c.execute(sthEntryExpType, (ExpTypes[expType],idCode))
            expTypesbyCode[idCode] = expType
print("ok")
##
##
expClasses = {}
compTypes = {}
sthExpClasse = 'INSERT INTO expClasse (expClasse) VALUES (%s)'
sthcompType = 'INSERT INTO comptype (type) VALUES (%s)'
entryUpdate = 'UPDATE entry SET idCompType=%s WHERE idCode=%s'
expTypeUpdate = 'UPDATE expType SET idExpClasse=%s where ExpType=%s'
with open('pdb_entry_type.txt','r') as EXPCL:
    for line in EXPCL:
        line = line.rstrip()
        idCode, compType, expClass = line.split()
        idCode = idCode.upper()
        with connection.cursor() as c:
            if expClass not in expClasses:
                c.execute(sthExpClasse, expClass)
                expClasses[expClass] = c.lastrowid
            if compType not in compTypes:
                c.execute(sthcompType, compType)
                compTypes[compType] = c.lastrowid
            c.execute(expTypeUpdate, (expClasses[expClass], expTypesbyCode[idCode]))
            c.execute(entryUpdate, (compTypes[compType],idCode))
##Sequence
print("Sequences...")
#codes = []
#CODES = {}
sthInsert= "INSERT INTO sequence (idCode,chain,sequence,header) VALUES (%s,%s,%s,%s)"
seq = ''
idPdb = ''
chain = ''
header = ''
header_re = re.compile(r'^>([^_]*)_(.*)mol:(\S*) length:(\S*)')
with open("pdb_seqres.txt", 'r') as SEQS: 
    for line in SEQS:
        line =line.rstrip()
        if '>' == line[0]:
            if seq:
                seq = seq.replace("\n","")
                idPdb = idPdb.upper()
                chain = chain.replace(' ','')
                connection.cursor().execute(sthInsert,(idPdb,chain,seq, header))
                seq = ''
            groups = header_re.match(line)
            idPdb = groups.group(1)
            chain = groups.group(2)
            header = line.replace('>','')
        else:
            seq += line
print("ok")
connection.cursor().execute("SET FOREIGN_KEY_CHECKS=1")
